### DISCUSSION TITLE
How to define route templates containing `$` in URL paths?

### DISCUSSION THREAD

**Discussion by kvskranthikumar** [ThreeMammals/Ocelot#1930] [Q&A] [answered]
I have this configuration in ReRoutes -

```json
{
            "DownstreamPathTemplate": "/Users/api/v1/odata/$metadata",
            "DownstreamScheme": "http",
            "DownstreamHostAndPorts": [
                {
                    "Host": "localhost",
                    "Port": 7004
                }
            ],
            "UpstreamPathTemplate": "/Users/metadata",
            "UpstreamHttpMethod": [ "Get", "Post", "Delete" ]
 }
```
Which works perfectly, but I have a requirement where the **UpstreamPathTemplate** is `/Users/$metadata`

I observed that if the **UpstreamPathTemplate** has a dollar `$` then we get an error ->>
```
Ocelot.Responder.Middleware.ResponderMiddleware:
Warning: requestId: 800000b2-0003-f500-b63f-84710c7967bb, previousRequestId: no previous request id, 
message: 
  Error Code: UnableToFindDownstreamRouteError 
  Message: Failed to match ReRoute configuration for upstream path: /Users/$metadata, verb: GET. errors found in ResponderMiddleware. 
  Setting error response for request path:/Users/$metadata, request method: GET
```

**Comment by fvidesf** [comment]
how did you manage to solve this? I had the same problem. You would declare the UpstreamPathTemplate like "/\\$metadata", like you where escaping the $ character. The reason for this is I think is because internally Ocelota uses regex and $ gets interpreted different

**Comment by MJeorrett** [comment]
+1

I am also working with Ocelot and OData and trying to create an aggregated /$metadata endpoint in ocelot.  There are a few problems to solve and I have just come across this one.  Would love to know if there is a work around?

Thanks

Matthew

**Comment by MJeorrett** [comment]
To answer my own question, you can use the escape sequence '[$]' e.g. '/odata/[$]metadata' in the UpstreamPathTemplate.

**Reply by raman-m** [reply, marked as answer]
Yes, it can be a solution.

But, better to define one route with correct templates via [Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#placeholders) feature like 👇 
```json
{
  "DownstreamPathTemplate": "/Users/api/v1/odata/{metadata}",
  "UpstreamPathTemplate": "/Users/{metadata}",
}
```
